rocksdb/transactiondb.nim

```diff
@@ -0,0 +1,101 @@
+# Nim-RocksDB
+# Copyright 2024 Status Research & Development GmbH
+# Licensed under either of
+#
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
+#  * GPL license, version 2.0, ([LICENSE-GPLv2](LICENSE-GPLv2) or https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)
+#
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
+{.push raises: [].}
+
+import
+  std/sequtils,
+  ./lib/librocksdb,
+  ./options/[dbopts, readopts, writeopts],
+  ./transactions/[transaction, txdbopts, txopts],
+  ./columnfamily/[cfopts, cfdescriptor, cfhandle],
+  ./internal/[cftable, utils],
+  ./rocksresult
+
+export
+  dbopts,
+  txdbopts,
+  cfdescriptor,
+  readopts,
+  writeopts,
+  txopts,
+  transaction,
+  rocksresult
+
+type
+  TransactionDbPtr* = ptr rocksdb_transactiondb_t
+
+  TransactionDbRef* = ref object
+    cPtr: TransactionDbPtr
+    path: string
+    dbOpts: DbOptionsRef
+    txDbOpts: TransactionDbOptionsRef
+    cfTable: ColFamilyTableRef
+
+proc openTransactionDb*(
+    path: string,
+    dbOpts = defaultDbOptions(),
+    txDbOpts = defaultTransactionDbOptions(),
+    columnFamilies = @[defaultColFamilyDescriptor()]): RocksDBResult[TransactionDbRef] =
+
+  if columnFamilies.len == 0:
+    return err("rocksdb: no column families")
+
+  var
+    cfNames = columnFamilies.mapIt(it.name().cstring)
+    cfOpts = columnFamilies.mapIt(it.options.cPtr)
+    columnFamilyHandles = newSeq[ColFamilyHandlePtr](columnFamilies.len)
+    errors: cstring
+
+  let txDbPtr = rocksdb_transactiondb_open_column_families(
+        dbOpts.cPtr,
+        txDbOpts.cPtr,
+        path.cstring,
+        cfNames.len().cint,
+        cast[cstringArray](cfNames[0].addr),
+        cfOpts[0].addr,
+        columnFamilyHandles[0].addr,
+        cast[cstringArray](errors.addr))
```

Comment:
`cast[cstringArray](errors.addr))` although the generated code is correct, but conceptually, the `errptr` is not an array but an output pointer to `cstring`. Better to use `cast[ptr cstring](...)` to avoid confusion.
The C FFI should be fixed too.

Or better, we can use Nim `var` modifier, and we can avoid cast altogether.

Example:

```Nim
proc rocksdb_open*(options: ptr rocksdb_options_t; name: cstring; errptr: var cstring): ptr rocksdb_t 

var error: cstring
var rocksdb_open(opts, "name".cstring, error)
```

Reply:
Well done (I only looked at formalities, though.) Some remarks:
* I did not see bulk/sst support (maybe overlooked?)
* Important: Documentation for the API is missing, There should be either references or manually added documentation so that module can be easily applied.
* Transaction mode for rocks-db differs from Legacy/Aristo transactions in `nimbus-eth1`. So it would not easily replace the key-value tables at the moment. Some documentation would be nice :).

Reply:
> `cast[cstringArray](errors.addr))` although the generated code is correct, but conceptually, the `errptr` is not an array but an output pointer to `cstring`. Better to use `cast[ptr cstring](...)` to avoid confusion. The C FFI should be fixed too.
> 
> Or better, we can use Nim `var` modifier, and we can avoid cast altogether.
> 
> Example:
> 
> ```nim
> proc rocksdb_open*(options: ptr rocksdb_options_t; name: cstring; errptr: var cstring): ptr rocksdb_t 
> 
> var error: cstring
> var rocksdb_open(opts, "name".cstring, error)
> ```

@jangko I tried what you suggested to use the cast to ptr cstring but it won't compile. For that to work I would need to change the nim c wrapper code. I want to avoid manually changing the c wrapper though because it was generated by the c2nim tool and there are many places to update.  If we want to update the library again in the future we would need to go through the whole process again so I would suggest that the minor readability improvement may not be worth the effort in this case.